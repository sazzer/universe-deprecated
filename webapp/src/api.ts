import axios, { Method, AxiosResponse } from 'axios';
import * as rfc6570 from 'rfc6570-expand';

declare global {
  interface Window {
    _UNIVERSE_CONFIG: any;
  }
}

/**
 * The shape of the request that we want to make
 */
export interface Request {
  url: string;
  urlParams?: any;
  method?: Method;
  headers?: any;
  data?: any;
}

export type Response<T> = AxiosResponse<T>;

/**
 * Shape of a Problem response from the serve
 */
export interface Problem {
  type: string,
  status: number,
}

/**
 * Response that is thrown if we get an RFC-7807 Problem back from the server
 */
export class ProblemResponse<T extends Problem> {
  readonly problem: T;
  readonly status: number;
  readonly headers: any;

  constructor(problem: T, status: number, headers: any) {
    this.problem = problem;
    this.status = status;
    this.headers = headers;
  }
}

/**
 * Build an API requester to make API calls with
 */
export async function request<T>(request: Request): Promise<Response<T>> {
  const serviceUrl = process.env.REACT_APP_SERVICE_URL || ((window || {})._UNIVERSE_CONFIG || {}).serviceUrl;

  const template = rfc6570.init(request.url);
  const expandedUri = template.expand(request.urlParams || {});
  try {
    return await axios.request({
      baseURL: serviceUrl,
      timeout: 20000,
      url: expandedUri,
      method: request.method,
      headers: request.headers,
      data: request.data,
    });
  } catch (e) {
    if (e.response) {
      const response = e.response as AxiosResponse<Problem>;
      if (response.headers['content-type'] === 'application/problem+json') {
        throw new ProblemResponse(response.data, response.status, response.headers);
      }
    }
    console.error('Unexpected error making API request', e);
    throw e;
  }
}
