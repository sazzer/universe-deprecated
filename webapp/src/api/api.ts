import * as rfc6570 from "rfc6570-expand";

import { Problem, ProblemResponse } from "./problem";
import axios, { AxiosResponse, Method } from "axios";

import debug from "debug";

/** The logger to use */
const LOG = debug("universe:api");

/** The error logger to use */
const ERROR_LOG = debug("universe:api:error");
debug.enable("universe:api:error");

declare global {
  interface Window {
    _UNIVERSE_CONFIG: any;
  }
}

/** The access token to use */
let _accessToken: string | undefined;

/**
 * The shape of the request that we want to make
 */
export interface Request {
  url: string;
  urlParams?: any;
  method?: Method;
  headers?: any;
  data?: any;
}

/**
 * The shape of a Response from a request
 */
export type Response<T> = AxiosResponse<T>;

/**
 * Build an API requester to make API calls with
 */
export async function request<T>(request: Request): Promise<Response<T>> {
  let serviceUrl = process.env.REACT_APP_SERVICE_URL;
  if (window && window._UNIVERSE_CONFIG && window._UNIVERSE_CONFIG.serviceUrl) {
    serviceUrl = window._UNIVERSE_CONFIG.serviceUrl;
  }

  const template = rfc6570.init(request.url);
  const expandedUri = template.expand(request.urlParams || {});
  LOG("Making request to: %s", expandedUri);

  const headers = {
    ...request.headers
  };

  if (_accessToken) {
    headers.Authorization = `Bearer ${_accessToken}`;
  }

  try {
    return await axios.request({
      baseURL: serviceUrl,
      timeout: 20000,
      url: expandedUri,
      method: request.method,
      headers: headers,
      data: request.data
    });
  } catch (e) {
    if (e.response) {
      const response = e.response as AxiosResponse<Problem>;
      if (response.headers["content-type"] === "application/problem+json") {
        LOG("Received an RFC-7807 Problem response: %o", response.data);
        throw new ProblemResponse(
          response.data,
          response.status,
          response.headers
        );
      }
    }

    ERROR_LOG("Unexpected error making API request", e);
    throw e;
  }
}

/**
 * Set the access token that we are going to use for future requests
 * @param accessToken The access token, or `undefined` to clear it
 */
export function setAccessToken(accessToken: string | undefined) {
  _accessToken = accessToken;
}
