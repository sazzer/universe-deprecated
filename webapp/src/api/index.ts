export * from "./api";
export * from "./problem";
export * from "./validation";
