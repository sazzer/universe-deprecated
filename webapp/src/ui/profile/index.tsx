import React from "react";
import { Link, useRouteMatch, NavLink, Switch, Route } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { LoggedIn } from "../login/loggedIn";
import { UserProfileArea } from "./profile";

/** Props to represent an entry in the profile page */
interface ProfilePageEntry {
  name: string;
  path: string;
  label: string;
  content: any;
}

/** Component to represent a single tab on the profile page */
const ProfilePageTab: React.FC<ProfilePageEntry> = ({ path, name, label }) => {
  const match = useRouteMatch({ path: path, exact: true });

  return (
    <NavLink
      to={path}
      exact
      className="nav-link"
      id={`profile-tabs-${name}-tab`}
      role="tab"
      aria-controls={`profile-tabs-${name}`}
      aria-selected={match !== null}
    >
      {label}
    </NavLink>
  );
};

/** Component to represent the actual body of the profile page */
const ProfilePagePane: React.FC<ProfilePageEntry> = ({
  path,
  name,
  content
}) => {
  return (
    <Route exact path={path}>
      <div
        aria-labelledby={`profile-tabs-${name}-tab`}
        role="tabpanel"
        id={`profile-tabs-${name}`}
      >
        {content}
      </div>
    </Route>
  );
};

/**
 * React Component represnting the user profile page
 */
export const ProfilePage: React.FC = () => {
  const { t } = useTranslation();

  const pages = [
    {
      name: "profile",
      path: "/profile",
      label: t("profile.profile.label"),
      content: <UserProfileArea />
    },
    {
      name: "worlds",
      path: "/profile/worlds",
      label: t("profile.worlds.label"),
      content: "My Worlds Content"
    },
    {
      name: "stories",
      path: "/profile/stories",
      label: t("profile.stories.label"),
      content: "My Stories Content"
    }
  ];

  const tabs = pages.map(page => (
    <ProfilePageTab
      key={page.name}
      path={page.path}
      name={page.name}
      label={page.label}
      content={page.content}
    />
  ));

  const panes = pages
    .reverse()
    .map(page => (
      <ProfilePagePane
        key={page.name}
        path={page.path}
        name={page.name}
        label={page.label}
        content={page.content}
      />
    ));

  return (
    <LoggedIn>
      <nav aria-label="breadcrumb">
        <ol className="breadcrumb">
          <li className="breadcrumb-item">
            <Link to="/">{t("profile.breadcrumbs.home")}</Link>
          </li>
          <li className="breadcrumb-item active" aria-current="page">
            {t("profile.breadcrumbs.profile")}
          </li>
        </ol>
      </nav>

      <div className="row">
        <div className="col-12 col-md-9 order-sm-1">
          <div className="tab-content">
            <Switch>{panes}</Switch>
          </div>
        </div>

        <div className="col-12 col-md-3">
          <div
            className="nav flex-column nav-pills"
            id="profileSupportedContent"
            role="tablist"
            aria-orientation="vertical"
          >
            {tabs}
          </div>
        </div>
      </div>
    </LoggedIn>
  );
};
