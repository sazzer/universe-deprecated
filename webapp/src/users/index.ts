export * from "./model";
export * from "./userContext";
export * from "./api";
