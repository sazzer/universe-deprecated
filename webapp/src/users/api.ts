import { ProblemResponse, request } from "../api";

import { AuthenticatedUser } from "./model";
import debug from "debug";

/** The logger to use */
const LOG = debug("universe:users:api");

/**
 * Check if a given username is registered with the service or not
 * @param username The username to check
 * @return True if the username is already registered. False if it's not.
 */
export async function checkUsername(username: string): Promise<boolean> {
  LOG("Checking the existance of username: %s", username);
  try {
    await request({
      url: "/usernames/{username}",
      urlParams: {
        username
      },
      method: "GET"
    });

    LOG("Existance of username %s: true", username);
    return true;
  } catch (e) {
    if (
      e instanceof ProblemResponse &&
      e.problem.type === "tag:universe,2020:users/problems/unknown-user"
    ) {
      LOG("Existance of username %s: false", username);
      return false;
    } else {
      LOG("Error checking existance of username %s: %o", username, e);
      throw e;
    }
  }
}

/**
 * Attempt to authenticate with the given credentials
 * @param username the username to authenticate with
 * @param password the password to authenticate with
 */
export async function authenticate(
  username: string,
  password: string
): Promise<AuthenticatedUser> {
  LOG("Authenticating as username %s with password %s", username, password);
  try {
    const user = await request<AuthenticatedUser>({
      url: "/login",
      method: "POST",
      data: {
        username,
        password
      }
    });

    LOG("Authenticated successfully: %o", user);
    return user.data;
  } catch (e) {
    LOG("Failed to authenticate as username %s: %o", username, e);
    throw e;
  }
}
