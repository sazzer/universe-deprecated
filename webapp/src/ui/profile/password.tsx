import * as yup from "yup";

import { ErrorMessage, FieldValues, useForm } from "react-hook-form";
import React, { useState } from "react";
import { changePassword, useUser } from "../../users";

import { SubmitButton } from "../components/form/buttons";
import { UnexpectedError } from "../components/form/error";
import { ValidationErrors } from "../../api";
import debug from "debug";
import { useTranslation } from "react-i18next";

/** The logger to use */
const LOG = debug("universe:ui:profile:changePassword");

/**
 * The actual Change Password form to work with
 */
export const ChangePasswordArea: React.FC = () => {
  const { t } = useTranslation();
  const [loading, setLoading] = useState(false);
  const [globalError, setGlobalError] = useState<string | undefined>(undefined);
  const { user } = useUser();

  const { register, errors, handleSubmit, setError } = useForm({
    validationSchema: yup.object().shape({
      password: yup
        .string()
        .required(
          t(
            "profile.password.password.errors.tag:universe,2020:validation-errors/missing"
          )
        ),
      password2: yup
        .string()
        .required(
          t(
            "profile.password.password2.errors.tag:universe,2020:validation-errors/missing"
          )
        )
        .when(["password"], (password: string, schema: any) => {
          return schema.oneOf(
            [password],
            t(
              "profile.password.password2.errors.tag:universe,2020:validation-errors/password/invalid-password"
            )
          );
        })
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      password: "",
      password2: ""
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    LOG("Submitting form: %o", data);
    setGlobalError(undefined);
    setLoading(true);

    try {
      const saved = await changePassword(user?.id || "", data.password);
    } catch (e) {
      if (e instanceof ValidationErrors) {
        e.errors.forEach(error => {
          const message = t(
            `profile.profile.${error.field}.errors.${error.type}`
          );
          setError(error.field, error.type, message);
        });
      } else {
        setGlobalError(e.toString());
      }
    }
    setLoading(false);
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmitHandler)}
      data-test="changePasswordForm"
    >
      <div className="form-group" data-test="password">
        <label htmlFor="password">{t("profile.password.password.label")}</label>
        <input
          type="password"
          className={
            errors.password ? "form-control is-invalid" : "form-control"
          }
          id="password"
          name="password"
          readOnly={loading}
          autoFocus
          ref={register}
        />
        <ErrorMessage
          errors={errors}
          name="password"
          as={<div className="invalid-feedback" />}
        />
      </div>
      <div className="form-group" data-test="password2">
        <label htmlFor="password2">
          {t("profile.password.password2.label")}
        </label>
        <input
          type="password"
          className={
            errors.password2 ? "form-control is-invalid" : "form-control"
          }
          id="password2"
          name="password2"
          readOnly={loading}
          ref={register}
        />
        <ErrorMessage
          errors={errors}
          name="password2"
          as={<div className="invalid-feedback" />}
        />
      </div>
      <div className="form-group">
        <SubmitButton loading={loading}>
          {t("profile.password.submit")}
        </SubmitButton>
      </div>
      {globalError && <UnexpectedError message={globalError} />}
    </form>
  );
};
