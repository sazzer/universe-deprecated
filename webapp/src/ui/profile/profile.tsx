import React, { useEffect } from "react";
import { useForm, ErrorMessage, FieldValues } from "react-hook-form";
import { useTranslation } from "react-i18next";
import * as yup from "yup";
import { useOvermind } from "../../overmind";
import { Loader } from "../loader";
import { User, UserState } from "../../overmind/users/model";

/**
 * Props for rendering the User Profile Form
 */
interface UserProfileFormProps {
  user: User;
  userState: UserState;
}

/**
 * React component to actually render the user profile form
 */
export const UserProfileForm: React.FC<UserProfileFormProps> = ({ user }) => {
  const { t } = useTranslation();

  const { register, errors, handleSubmit } = useForm({
    validationSchema: yup.object().shape({
      username: yup
        .string()
        .required(
          t(
            "profile.profile.username.errors.tag:universe,2020:validation-errors/missing"
          )
        )
        .trim(),
      email: yup
        .string()
        .email(
          t(
            "login.email.errors.tag:universe,2020:users/validation-errors/email/malformed"
          )
        )
        .required(
          t(
            "profile.profile.email.errors.tag:universe,2020:validation-errors/missing"
          )
        )
        .trim(),
      displayName: yup
        .string()
        .required(
          t(
            "login.displayName.errors.tag:universe,2020:validation-errors/missing"
          )
        )
        .trim()
    }),
    validateCriteriaMode: "all",
    defaultValues: {
      username: user.username,
      email: user.email || "",
      displayName: user.displayName
    }
  });

  const onSubmitHandler = async (data: FieldValues) => {
    console.log(data);
  };

  return (
    <>
      <form
        onSubmit={handleSubmit(onSubmitHandler)}
        data-test="userProfileForm"
      >
        <div className="form-group" data-test="username">
          <label htmlFor="username">
            {t("profile.profile.username.label")}
          </label>
          <input
            type="text"
            className={
              errors.username ? "form-control is-invalid" : "form-control"
            }
            id="username"
            name="username"
            readOnly
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="username"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="email">
          <label htmlFor="email">{t("profile.profile.email.label")}</label>
          <input
            type="text"
            className={
              errors.email ? "form-control is-invalid" : "form-control"
            }
            id="email"
            name="email"
            autoFocus
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="email"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group" data-test="displayName">
          <label htmlFor="displayName">
            {t("profile.profile.displayName.label")}
          </label>
          <input
            type="text"
            className={
              errors.displayName ? "form-control is-invalid" : "form-control"
            }
            id="displayName"
            name="displayName"
            ref={register}
          />
          <ErrorMessage
            errors={errors}
            name="displayName"
            as={<div className="invalid-feedback" />}
          />
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-primary">
            {t("profile.profile.submit")}
          </button>
        </div>
      </form>
    </>
  );
};

/**
 * React Component to represent the user profile area of the profile page
 */
export const UserProfileArea: React.FC = () => {
  const { state, actions } = useOvermind();

  const currentUser = state.authentication.userId;

  useEffect(() => {
    if (currentUser !== null) {
      actions.users.fetchUser(currentUser);
    }
  }, [currentUser, actions.users]);

  const storedUser = state.users.userById(currentUser || "");
  const user = storedUser.user;

  if (storedUser.state === "LOADING" || user === undefined) {
    return <Loader />;
  } else {
    return <UserProfileForm userState={storedUser.state} user={user} />;
  }
};
