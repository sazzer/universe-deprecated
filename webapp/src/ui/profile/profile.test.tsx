import { getUserById, useUser, updateUserProfile } from "../../users";
import { render, wait, fireEvent } from "@testing-library/react";

import { UserProfileArea } from "./profile";
import React from "react";
import { ValidationErrors } from "../../api";

jest.mock("../../users");

const useUserMock = useUser as jest.Mock;
const getUserByIdMock = getUserById as jest.Mock;
const updateUserProfileMock = updateUserProfile as jest.Mock;

afterEach(() => {
  jest.resetAllMocks();
});

test("Rendering before user loaded", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockReturnValueOnce(new Promise(() => {}));

  const { container } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(0);
  expect(updateUserProfileMock).toBeCalledTimes(0);
});

test("Rendering after user loaded", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  const { container } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  expect(container).toMatchSnapshot();

  expect(storeUser).toBeCalledTimes(1);
  // Called in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(0);
});

test("Saving invalid email address", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  const { container, getByText, getByLabelText } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("Email Address"), {
      target: { value: "testuser@examplecom" }
    });
    fireEvent.click(getByText("Save Changes", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();

  expect(storeUser).toBeCalledTimes(1);
  // Called in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(0);
});

test("Saving blank fields", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  const { container, getByText, getByLabelText } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("Email Address"), {
      target: { value: "" }
    });
    fireEvent.change(getByLabelText("Display Name"), {
      target: { value: "" }
    });
    fireEvent.click(getByText("Save Changes", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();

  expect(storeUser).toBeCalledTimes(1);
  // Called in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(0);
});

test("Saving successfully", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  updateUserProfileMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "newuser@example.com",
    username: "testuser",
    displayName: "New User"
  });

  const { container, getByText, getByLabelText } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("Email Address"), {
      target: { value: "newuser@example.com" }
    });
    fireEvent.change(getByLabelText("Display Name"), {
      target: { value: "New User" }
    });
    fireEvent.click(getByText("Save Changes", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(2);
  // First is in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  // Second is in response to saving the profile
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "newuser@example.com",
    username: "testuser",
    displayName: "New User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(1);
  expect(updateUserProfileMock).toBeCalledWith(
    "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    "newuser@example.com",
    "New User"
  );
});

test("Saving a duplicate email address", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  updateUserProfileMock.mockRejectedValueOnce(
    new ValidationErrors([
      {
        field: "email",
        type: "tag:universe,2020:users/validation-errors/email/duplicate"
      }
    ])
  );

  const { container, getByText, getByLabelText } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("Email Address"), {
      target: { value: "newuser@example.com" }
    });
    fireEvent.change(getByLabelText("Display Name"), {
      target: { value: "New User" }
    });
    fireEvent.click(getByText("Save Changes", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(1);
  // Called in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(1);
  expect(updateUserProfileMock).toBeCalledWith(
    "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    "newuser@example.com",
    "New User"
  );
});

test("Saving with an unexpected error", async () => {
  const storeUser = jest.fn();

  useUserMock.mockReturnValue({
    hasUser: true,
    user: {
      id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
      email: "testuser@example.com",
      username: "testuser",
      displayName: "Test User"
    },
    clearUser: () => {},
    storeUser
  });

  getUserByIdMock.mockResolvedValueOnce({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });

  updateUserProfileMock.mockRejectedValueOnce(new Error("Network error"));

  const { container, getByText, getByLabelText } = render(<UserProfileArea />);

  // Waiting on getUserByIdMock
  await wait(() => {});

  await wait(() => {
    fireEvent.change(getByLabelText("Email Address"), {
      target: { value: "newuser@example.com" }
    });
    fireEvent.change(getByLabelText("Display Name"), {
      target: { value: "New User" }
    });
    fireEvent.click(getByText("Save Changes", { selector: "button" }));
  });

  expect(container).toMatchSnapshot();
  expect(storeUser).toBeCalledTimes(1);
  // Called in response to loading the user
  expect(storeUser).toBeCalledWith({
    id: "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    email: "testuser@example.com",
    username: "testuser",
    displayName: "Test User"
  });
  expect(updateUserProfileMock).toBeCalledTimes(1);
  expect(updateUserProfileMock).toBeCalledWith(
    "57c33107-b43e-4b53-a967-3ff89ccaaf16",
    "newuser@example.com",
    "New User"
  );
});
