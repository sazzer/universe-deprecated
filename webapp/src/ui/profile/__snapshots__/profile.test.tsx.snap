// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering after user loaded 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
  </form>
</div>
`;

exports[`Rendering before user loaded 1`] = `
<div>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Saving a duplicate email address 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Email Address is registered to another user
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving blank fields 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter an email address
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control is-invalid"
        id="displayName"
        name="displayName"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a display name
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving invalid email address 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control is-invalid"
        id="email"
        name="email"
        type="text"
      />
      <div
        class="invalid-feedback"
      >
        Please enter a valid email address
      </div>
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
  </form>
</div>
`;

exports[`Saving successfully 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-success"
        role="alert"
      >
        Password changed successfully
      </div>
    </div>
  </form>
</div>
`;

exports[`Saving with an unexpected error 1`] = `
<div>
  <form
    data-test="userProfileForm"
  >
    <div
      class="form-group"
      data-test="username"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        class="form-control"
        id="username"
        name="username"
        readonly=""
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="email"
    >
      <label
        for="email"
      >
        Email Address
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="form-group"
      data-test="displayName"
    >
      <label
        for="displayName"
      >
        Display Name
      </label>
      <input
        class="form-control"
        id="displayName"
        name="displayName"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Changes
      </button>
    </div>
    <div
      class="form-group"
    >
      <div
        class="alert alert-danger"
        role="alert"
      >
        An unexpected error occurred: Error: Network error
      </div>
    </div>
  </form>
</div>
`;
